webui/react/src/components/PaginatedNotesCard.tsx

```diff
@@ -0,0 +1,203 @@
+    if (pageNumber === currentPage) return;
+    if (editedContents !== notes?.[currentPage]?.contents) {
+      modal.confirm({
+        content: (
+          <p>
+            You have unsaved notes, are you sure you want to switch pages?
+            Unsaved notes will be lost.
+          </p>),
+        onOk: () => { setCurrentPage(pageNumber as number); fireNoteChangeSignal(); },
+        title: 'Unsaved content',
+      });
+    } else {
+      setCurrentPage(pageNumber as number);
+      setEditedContents(notes?.[currentPage]?.contents ?? '');
+      fireNoteChangeSignal();
+    }
+  }, [ currentPage, editedContents, modal, notes, fireNoteChangeSignal ]);
+
+  useEffect(() => {
+    if (previousNumberOfNotes == null) {
+      if (notes.length) {
+        handleSwitchPage(0);
+        fireNoteChangeSignal();
+      }
+    } else if (notes.length > previousNumberOfNotes) {
+      handleSwitchPage(notes.length - 1);
+    } else if (notes.length < previousNumberOfNotes){
+      // dont call handler here because page isn't actually switching
+      setCurrentPage((prevPageNumber) =>
+        prevPageNumber > deleteTarget ? prevPageNumber - 1 : prevPageNumber);
+    }
+  }, [ previousNumberOfNotes, notes.length, deleteTarget, handleSwitchPage, fireNoteChangeSignal ]);
+
+  const handleNewPage = useCallback(() => {
+    const currentPages = notes.length;
+    onNewPage();
+    handleSwitchPage(currentPages);
+  }, [ notes.length, onNewPage, handleSwitchPage ]);
+
+  const handleSave = useCallback(async (editedNotes: string) => {
+    setEditedContents(editedNotes);
+    await onSave(notes.map((note, idx) => {
+      if (idx === currentPage) {
+        return { contents: editedNotes, name: editedName } as Note;
+      }
+      return note;
+    }));
+  }, [ currentPage, editedName, notes, onSave ]);
+
+  const handleSaveTitle = useCallback(async (newName: string) => {
+    setEditedName(newName);
+    await onSave(notes.map((note, idx) => {
+      if (idx === currentPage) {
+        return { contents: editedContents ?? note?.contents, name: newName } as Note;
+      }
+      return note;
+    }));
+  }, [ currentPage, notes, onSave, editedContents ]);
+
+  const handleDeletePage = useCallback((deletePageNumber: number) => {
+    onDelete(deletePageNumber);
+    setDeleteTarget(deletePageNumber);
+  }, [ onDelete, setDeleteTarget ]);
+
+  const handleEditedNotes = useCallback((newContents: string) => {
+    setEditedContents(newContents);
+  }, []);
+
+  useEffect(() => {
+    if (notes.length === 0) return;
+    if (currentPage < 0) { setCurrentPage(0); fireNoteChangeSignal(); }
+    if (currentPage >= notes.length) { setCurrentPage(notes.length - 1); fireNoteChangeSignal(); }
+  }, [ currentPage, notes.length, fireNoteChangeSignal ]);
+
+  useEffect(() => {
+    setEditedContents(notes?.[currentPage]?.contents ?? '');
+    setEditedName(notes?.[currentPage]?.name ?? '');
```

Comment:
**question:** would it be better to not make any changes if the input is invalid? for example:
```
setEditedContents(prev => notes?.[currentPage]?.contents ?? prev);
```